### PR TITLE
Implement GSP-134 & GSP-654

### PR DESCRIPTION
**Pull request by JinnyYi**

### CODE REVIEW

File: storage.go [hunk 1]
```diff
@@ -76,34 +76,37 @@ func (s *Storage) create(path string, opt pairStorageCreate) (o *Object) {
...
+		o.SetContentType(opt.ContentType)
+	}
+	var sm ObjectSystemMetadata
+	if opt.HasStorageClass {
+		sm.StorageClass = opt.StorageClass
+	}
+	if opt.HasServerSideEncryption {
+		sm.ServerSideEncryption = opt.ServerSideEncryption
+	}
+
```
Comment (xxchan) [inline storage.go:109]:
```suggestion
	o.SetSystemMetadata(sm)

```

Is this missing?

File: service.toml [hunk 2]
```diff
@@ -36,7 +36,7 @@ optional = ["offset", "io_callback", "size"]
 optional = ["content_md5", "content_type", "io_callback", "storage_class", "server_side_encryption", "server_side_data_encryption", "server_side_encryption_key_id"]
 
 [namespace.storage.op.create_append]
-optional = ["content_type", "server_side_encryption", "server_side_encryption_key_id", "server_side_data_encryption", "storage_class"]
```
Comment (xxchan) [inline service.toml:39]: Why are they removed?

Comment (JinnyYi) [reply service.toml:39]: According to [aliyun API reference](https://help.aliyun.com/document_detail/31981.html), apend objects stored in the OSS server cannot be encrypted by using a specified CMK ID. And the request headers of `AppendObject` do not include `x-oss-server-side-data-encryption` and `x-oss-server-side-encryption-key-id`. So I removed the pairs.

File: storage.go [hunk 3]
```diff
@@ -76,34 +76,38 @@ func (s *Storage) create(path string, opt pairStorageCreate) (o *Object) {
...
-		options = append(options, oss.ServerSideEncryption(opt.ServerSideEncryption))
-	}
-	if opt.HasServerSideDataEncryption {
-		options = append(options, oss.ServerSideDataEncryption(opt.ServerSideDataEncryption))
-	}
-	if opt.HasServerSideEncryptionKeyID {
-		options = append(options, oss.ServerSideEncryptionKeyID(opt.ServerSideEncryptionKeyID))
-	}
-
-	offset, err := s.bucket.AppendObject(rp, nil, 0, options...)
```
Comment (xxchan) [inline storage.go:97]: This API call is removed (so won't create an empty object?). Is this expected?

Comment (xxchan) [reply storage.go:97]: similar question here https://github.com/beyondstorage/go-service-qingstor/pull/71/files#diff-8127f6766ba5110c7a99d07305f11d18295fcb2cb08dddb105ab7d55898742c7L113

Comment (JinnyYi) [reply storage.go:97]: The original intention of creating an empty object in `CreateAppend` is to set the object system metadata at the first append request. You can refer [here](https://github.com/beyondstorage/go-service-qingstor/pull/40#discussion_r621213243) for more details. 
Actually there's no requirement to create an empty object for `CreateAppend`, so I removed the API call to reduce a request.

Comment (xxchan) [reply storage.go:97]: The discussion says that if we cannot create an empty object, we use metadata. 
You mean that we can simply use metadata? This LGTM, but I'm not sure if there is further consideration.

Comment (JinnyYi) [reply storage.go:97]: > You mean that we can simply use metadata? This LGTM, but I'm not sure if there is further consideration.

I'm not sure whether we should ensoure that the object can be retrieved from server after `CreateAppend`. Or just create an appendable `Object` for users to call `WriteAppend`?

Comment (Xuanwo) [reply storage.go:97]: > I'm not sure whether we should ensoure that the object can be retrieved from server after `CreateAppend`. Or just create an appendable `Object` for users to call `WriteAppend`?

I think it's not safe. For example, user call `CreateAppend()` but failed on next `WriteAppend` call. So, it's impossible for him to `Stat()` and resume the append process.

Comment (JinnyYi) [reply storage.go:97]: Got it. I'll change back to create an empty object in `CreateAppend`.

File: storage.go [hunk 4]
```diff
@@ -76,34 +76,55 @@ func (s *Storage) create(path string, opt pairStorageCreate) (o *Object) {
...
 	if err != nil {
 		return
 	}
 
+	if isExist {
+		err = s.bucket.DeleteObject(rp)
+		if err != nil {
+			return
+		}
+	} else {
```
Comment (xxchan) [inline storage.go:91]: We shouldn't have else here? (if exist, delete and then create)

**Review by xxchan — APPROVED** [review]

**Pull request merged**